### PR TITLE
Moving to Pluto 🎈

### PR DESCRIPTION
**Pull request by MichielStock**
Migrating all notebooks to Pluto

currently done:

- [x] introduction
- [x] quadratic
- [x] brackets
- [x] autodiff
- [x] unconstrained convex
- [x] constrained convex
- [x] optimal transport
- [x] mst
- [ ] shortest paths
- [ ] NP-complete
- [ ] metaheuristics
- [ ] TSP

Ping @kirstvh to check these

### CODE REVIEW

File: chapters/01.Brackets/brackets.jl [hunk 1]
```diff
@@ -0,0 +1,1156 @@
...
+"""
+
+# ╔═╡ 6b30fe20-f321-4cb8-a2fc-0a2820d71bc0
+md"As a first introduction to mathematical optimization, we will study two *bracketing methods*. Bracketing methods can be used to minimize scalar function with a single input variable. These algorithms identify an interval $[a, b]$ containing the desired minimum. We will assume that the functions are *unimodal* they contain a single minimum."
+
+# ╔═╡ bdf6c059-db72-4d90-b98f-2e77b7d89f37
+md"""
+## Finding an initial bracket
+
+The following function will generate from an initial $x$ an interval containing a local minimum. It works by moving one of the limits of the interval until the function value at that limit starts to decrease.
```
Comment (kirstvh) [inline chapters/01.Brackets/brackets.jl]: It works by moving one of the limits of the interval until the function value starts to **increase**?

File: chapters/01.Brackets/brackets.jl [hunk 2]
```diff
@@ -0,0 +1,1156 @@
...
+md"""
+# Bracket search
+
+*STMO*
+
+**Michiel Stock**
+"""
+
+# ╔═╡ 6b30fe20-f321-4cb8-a2fc-0a2820d71bc0
+md"As a first introduction to mathematical optimization, we will study two *bracketing methods*. Bracketing methods can be used to minimize scalar function with a single input variable. These algorithms identify an interval $[a, b]$ containing the desired minimum. We will assume that the functions are *unimodal* they contain a single minimum."
```
Comment (kirstvh) [inline chapters/01.Brackets/brackets.jl]: ✏️ We will assume that the functions are *unimodal* **, i.e.,**  they contain a single minimum.

File: chapters/01.Brackets/brackets.jl [hunk 3]
```diff
@@ -0,0 +1,1156 @@
...
+
+The bisection method is technically not a minimization algorithm, but a *root finding method*, i.e., it can be used to find an $x$ for a given function $g: \mathbb{R}\rightarrow \mathbb{R}$ such that
+
+$$g(x) = 0\,.$$
+
+We can use the bisection method to find a minimum of a function $f(x)$ by finding points where the first derivative $f'(x)$ is equal to 0.
+
+> **Question**: in addition to the first derivative being equal to zero, which other criteria should hold for $x^\star$ to be a minimizer?
+
+The bisection method departs from an initial bracket $[a, b]$, where $g(a)$ and $g(b)$ have opposing signs. If $g(x)$ is a continious function, the *intermediate value theorem* states that there is at least one $x\in[a,b]$ such that $g(x)$.
```
Comment (kirstvh) [inline chapters/01.Brackets/brackets.jl]: ✏️ If $g(x)$ is a **continuous** function, the intermediate value theorem states that there is at least one $x\in[a,b]$ such that $g(x)$ **= 0**.

File: chapters/01.Brackets/brackets.jl [hunk 4]
```diff
@@ -0,0 +1,1156 @@
...
+
+In every step of the bisection method, the interval it cut into half. The midpoint $x = (a+b)/2$ is computed, and a new bracket is formed from the midpoint and the side that contains zero.
```
Comment (kirstvh) [inline chapters/01.Brackets/brackets.jl]: ✏️ In every step of the bisection method, the interval **is** cut into half.

File: chapters/02.Quadratic/quadratic.jl [hunk 5]
```diff
@@ -0,0 +1,1968 @@
...
+
+$$\min_x\,\frac{1}{2}px^2+qx +r\,.$$
+
+This can easily be solved by setting the first order derivative equal to zero:
+
+$$\frac{\mathrm{d}f(x)}{\mathrm{d}x} = px + q$$
+
+$$px^\star+q = 0 \Leftrightarrow x^\star=\frac{-q}{p}\,.$$
+
+To show that this is the sole minimizer of $f(x)$, we have to prove that the second-order derivative is positive at this point. This means that at that point, the derivative of the function is increasing: a little to the left the function is increasing, a little to the right and the function is decreasing. We have
```
Comment (kirstvh) [inline chapters/02.Quadratic/quadratic.jl]: This means that at that point, the derivative of the function is increasing: a little to the left the function is **decreasing**, a little to the right and the function is **increasing**?

File: chapters/04.Unconstrained/unconstrained.jl [hunk 6]
```diff
@@ -0,0 +1,2207 @@
...
+If the first- and second order derivatives exists, a strongly $m$-convex function satisfies:
+
+-  $f(\mathbf{x}') \geq f(\mathbf{x}) + \nabla f(\mathbf{x})^\top (\mathbf{x}'-\mathbf{x}) + \frac{m}{2}||\mathbf{x}'-\mathbf{x}||_2$
+-  $\nabla^2 f(\mathbf{x})-mI\succeq 0$ (all eigenvalues of the Hessian are greater than $m$)
+
+If a function is $m$-strongly convex, this also implies that there exists an $M>m$ such that
+
+$$\nabla^2 f(\mathbf{x}) \preceq MI\,.$$
+
+Stated differently, for strongly convex functions the exist both a quadratic function with a smaller as well as a lower local curvature.
```
Comment (kirstvh) [inline chapters/04.Unconstrained/unconstrained.jl:306]: Stated differently, for strongly convex functions **there** exists both a quadratic function with a **higher** as well as a **lower** local curvature ?

File: chapters/04.Unconstrained/unconstrained.jl [hunk 7]
```diff
@@ -0,0 +1,2207 @@
...
+Convex functions are usually minimized using descent methods. Again, line search is often used as a subroutine.
+
+The outline of a general descent algorithm is given in the following pseudocode.
+
+> **input** starting point $\mathbf{x}\in$ **dom** $f$.
+>
+> **repeat**
+>
+>>    1. Determine a descent direction $\Delta \mathbf{x}$.
+>>    2. *Line seach*. Choose a step size $t>0$.
```
Comment (kirstvh) [inline chapters/04.Unconstrained/unconstrained.jl:337]: ✏️ Line **search**

File: chapters/04.Unconstrained/unconstrained.jl [hunk 8]
```diff
@@ -0,0 +1,2207 @@
...
+# ╔═╡ 723d9b31-eb75-49c1-a33a-9e105d3ccdfb
+md"""
+### General gradient descent algorithm
+
+>**input** starting point $\mathbf{x}\in$ **dom** $f$.
+>
+>**repeat**
+>
+>>    1. *Choose direction*. $\Delta \mathbf{x} := -\nabla f(\mathbf{x})$.
+>>    2. *Line seach*. Choose a step size $t$ via exact or backtracking line search.
```
Comment (kirstvh) [inline chapters/04.Unconstrained/unconstrained.jl:466]: ✏️ Line **search**

File: chapters/04.Unconstrained/unconstrained.jl [hunk 9]
```diff
@@ -0,0 +1,2207 @@
...
+
+Using the $L_1$ norm results in coordinate descent. For every iteration in this algorithm, we descent in the direction of the dimension where the absolute value of the gradient is largest.
+
+>**input** starting point $\mathbf{x}\in$ **dom** $f$.
+>
+>**repeat**
+>
+>>    1. *Direction*. Choose $i$ such that $|\nabla f(\mathbf{x})_i|$ is maximal.
+>>    2. *Choose direction*. $\Delta \mathbf{x} := -\nabla f(\mathbf{x})_i \mathbf{e}_i$
+>>    3. *Line seach*. Choose a step size $t$ via exact or backtracking line search.
```
Comment (kirstvh) [inline chapters/04.Unconstrained/unconstrained.jl:621]: ✏️ Line **search**

File: chapters/04.Unconstrained/unconstrained.jl [hunk 10]
```diff
@@ -0,0 +1,2207 @@
...
+
+### Pseudocode of Newton's algortihm
+
+>**input** starting point $\mathbf{x}\in$ **dom** $f$.
+>
+>**repeat**
+>
+>>    1. Compute the Newton step and decrement $\Delta \mathbf{x}_\text{nt} := -\nabla^2f(\mathbf{x})^{-1} \nabla f(\mathbf{x})$; $\lambda^2:=\nabla f(\mathbf{x})^\top\nabla^2 f(\mathbf{x})^{-1}\nabla f(\mathbf{x})$.
+>>    2. *Stopping criterion* **break** if $\lambda^2/2 \leq \epsilon$.
+>>    3. *Line seach*. Choose a step size $t$ via exact or backtracking line search.
```
Comment (kirstvh) [inline chapters/04.Unconstrained/unconstrained.jl:762]: ✏️ Line **search**

File: chapters/05.Constrained/constrained.jl [hunk 11]
```diff
@@ -0,0 +1,1644 @@
...
+**Michiel Stock**
+
+![](https://github.com/MichielStock/STMO/blob/master/chapters/05.Constrained/Figures/logo.png?raw=true)
+"""
+
+# ╔═╡ b3efad4a-51f0-4fba-9b35-df2e3914a76e
+md"""
+## Motivation
+
+Many more realistic optimization problems are characterized by constraints. For example, real-word systems often satisfy conservation laws, such as conservation of mass, of atoms or of charge. When designing objects, there are practical constraints of feasible dimensions, range of operations and limitations in materials. Another example is in probability, where a solution should satisfy the axioms of probability theory (probabilities are real values between 0 and 1 and the probabilities of all events should sum to 1).
```
Comment (kirstvh) [inline chapters/05.Constrained/constrained.jl:109]: ✏️ real-world

File: chapters/05.Constrained/constrained.jl [hunk 12]
```diff
@@ -0,0 +1,1644 @@
...
+In this chapter we discuss convex optimization problems with linear equality constraints (constraining the solution to a linear subspace) and convex inequality constrains (constraining the solution to convex subspace). Both types of constraints result again in a convex optimization problem.
+"""
+
+# ╔═╡ 34a331a9-4e2f-41ea-b75b-4dd17d9f0911
+md"""
+## Lagrange multipliers
+
+Lagrange multipliers are elegant ways of finding stationary points of a function of several variables given one or more constraints. We give a short introduction based on a geometric perspective.
+
+> IMPORTANT: most textbooks treat Lagrange multipliers from as maximization problems. Here they are treated as minimization problems to be consistent with other chapters.
```
Comment (kirstvh) [inline chapters/05.Constrained/constrained.jl:122]: ✏️ most textbooks treat Lagrange multipliers _/from/_ as maximization problems.

File: chapters/05.Constrained/constrained.jl [hunk 13]
```diff
@@ -0,0 +1,1644 @@
...
+$$\nabla f_0(\mathbf{x}^\star)+\sum_{i=1}^m\lambda^\star_i\nabla f_i(\mathbf{x}^\star) +A^\top \boldsymbol{\nu}^\star=0$$
+
+$$\lambda_if_i(\mathbf{x}^\star)=0, \quad i=1,\ldots,m\,.$$
+"""
+
+# ╔═╡ 1235659a-50a4-47b6-ad8e-cd20c8888a6e
+md"""
+### Implicit constraints
+
+Rather than solving a minimization problems with inequality constraints, we can reformulate the objective function to include only the feasible regions:
```
Comment (kirstvh) [inline chapters/05.Constrained/constrained.jl:466]: ✏️ **problem**

File: chapters/04.Unconstrained/unconstrained.jl (hunk 10 above)
Comment (MichielStock) [reply chapters/04.Unconstrained/unconstrained.jl:762]: d'oh!

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 14]
```diff
@@ -0,0 +1,1792 @@
...
+**Michiel Stock**
+
+![](https://github.com/MichielStock/STMO/blob/master/chapters/06.OptimalTransport/Figures/logo.png?raw=true)
+"""
+
+# ╔═╡ ccb9bf05-1830-4c56-bf28-312c5d324484
+md"""
+## Motivation
+
+Optimal transportation deals with transforming one distribution into another while minizing a cost function. The problem can be formulated as a maximum entropy function, leading to an elegant solution using the *Sinkhorn algorithm*. Optimal transportation is used in computer vision, machine learning, astonomy, bioinformatics, ecology etc.
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:74]: ✏️ minimizing

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 15]
```diff
@@ -0,0 +1,1792 @@
...
+
+# ╔═╡ 1468ec48-bedb-4977-8b88-794d89e5ae8f
+C_cells = missing
+
+# ╔═╡ 88569ef5-bc60-4a77-aa72-fc9bdf66b3a4
+# add the lines to the cells
+
+# ╔═╡ da2cc178-64c7-4b30-909a-24fb402a489b
+md"""
+## The optimal transportion problem
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:181]: ✏️ transportation

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 16]
```diff
@@ -0,0 +1,1792 @@
...
+
+Let us introduce some notation so we can formally state this as an optimization problem. Let $\mathbf{a}$ be the vector containing the amount of dessert every person may eat. In this case $\mathbf{a} = [3,3,3,4,2,2,2,1]^\intercal$ (in general the dimension of $\mathbf{r}$ is $n$). Similarly, $\mathbf{c}$ denotes the vector of how much there is of every dessert, i.e. $\mathbf{b}=[4, 2, 6, 4, 4]^\intercal$ (in general the dimension of $\mathbf{b}$ is $m$). Often $\mathbf{a}$ and $\mathbf{b}$ represent marginal probability distributions, hence their values are nonzero sum to one.
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:183]: ✏️Similarly, **b** denotes the vector of how much there is of every dessert

"hence their values are nonzero sum to one. " ?

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 17]
```diff
@@ -0,0 +1,1792 @@
...
+
+The preferences of each person for each dessert is also stored in a matrix. To be consistent with the literature, this will be stored in a $n\times m$ *cost* matrix $C$. The above matrix is a preference matrix that can easily be changed into a cost matrix by flipping the sign of every element.
+
+So finally, the problem we want to solve is formally posed as
+
+$$d_C(\mathbf{r}, \mathbf{c}) = \min_{P\in U(\mathbf{a}, \mathbf{b})}\, \sum_{i,j}P_{ij}C_{ij}\,.$$
+
+This is called the *optimal transportation* between $\mathbf{a}$ and $\mathbf{b}$. It can be solved relatively easily using linear programming.
+
+The optimum, $d_C(\mathbf{a}, \mathbf{b})$, is called the *Wasserstein metric*. It is basically a distance between two probability distributions. It is sometimes also called the *earth mover distance* as it can be interpreted as how much 'dirt' you have to move to change one 'landscape' (distribution) in another.
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:199]: It is basically **the minimum distance** between two probability distributions ?

Comment (MichielStock) [reply chapters/06.OptimalTransport/optimal_transportation.jl:199]: Sentence was correct!

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 18]
```diff
@@ -0,0 +1,1792 @@
...
+"""
+
+# ╔═╡ 531be0de-e969-42db-a2fa-f9d125797296
+md"""
+## Application: color transfer
+
+In this exercise, we will apply optimal transportation to transfer the color scheme of one image to the other. We will read two images and exact the colors of the pixels (RGB encoded). Using optimal transport, a softmatching can be obtained between the pixels of te repective images. Then, every pixel of the first image can be recomputed using a weighted sum of the pixels of the second image (or vice versa).
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:291]: ✏️ **extract** the colors of the pixels

✏️ **respective** images

File: chapters/06.OptimalTransport/optimal_transportation.jl [hunk 19]
```diff
@@ -0,0 +1,1792 @@
...
+colors2 = vec(image2ss)
+
+# ╔═╡ ad3eebaa-4fa2-438a-a088-16334dba3a89
+n_colors1 = length(colors1)
+
+# ╔═╡ 11deb453-bd6e-43b1-b1e1-61606644eef8
+n_colors2 = length(colors2)
+
+# ╔═╡ e52d67d0-b161-403a-85c6-1bb59354f2ee
+md"The `Colors` package contains a function `colordiff` that quantifies the perceptive difference between two colors. It it an ideal cost function!"
```
Comment (kirstvh) [inline chapters/06.OptimalTransport/optimal_transportation.jl:336]: ✏️ It **is** an ideal cost function!

File: chapters/07.MST/mst.jl [hunk 20]
```diff
@@ -0,0 +1,1527 @@
...
+## Ticket to ride
+
+As an illustration, we provide the graph of the famous boardgame *Ticket To Ride* (USA version). The goal of this game is to connect two cities on a map by placing a number of trains between them. Let's load the graph!
+"""
+
+# ╔═╡ d2c92e1d-a3cb-4070-8f2c-cf01e943ad45
+md"The weighted edges. The weight represents the connection cost."
+
+# ╔═╡ cda14c79-7a39-4ef5-8c59-1dd2f0fac202
+md"Let us plot this graph. We also have the coordinates of the cities in `cities_coordinates`. It is not needed to find the MST, but can help us make a draw the map."
```
Comment (kirstvh) [inline chapters/07.MST/mst.jl:399]: ✏️ can help us make _/a draw/_ the map.

File: README.md [hunk 21]
```diff
@@ -57,38 +58,49 @@ This package contains the solution to most implementation exercises in this cour
 
 ### Github
 
-Using Git or Github desktop is recommended for this course. In case you don't already have git or Github installed, this can be done by following the instructions for your operating system here [here](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) for Git and [here](https://desktop.github.com/) for GitHub desktop. Using git, clone (i.e., download the files of) the course repository by typing
+Using Git or Github desktop is can be helpful for this course. In case you don't already have Git or Github installed, this can be done by following the instructions for your operating system here [here](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git) for Git and [here](https://desktop.github.com/) for GitHub desktop. Using Git, clone (i.e., download the files of) the course repository by typing
```
Comment (kirstvh) [inline README.md:61]: ✏️ Using Git or Github desktop _is_ can be helpful for  ...

2 times _here_ [here](....)

File: README.md [hunk 22]
```diff
@@ -26,24 +26,25 @@ As of 2020, we have chosen to move this course from Python to the new [Julia pro
...
+ you can run the build file
 
 ```julia
 include("build.jl")
 ```
 
-Beware that building the course will take a while.
-
-In addition to the Jupyter notebooks and PDF notes, running this script also generates some example figures. We encourage students to take a look in the `scrips/` folder for some examples illustrating the theory. All PDF notes will also be made available on Ufora.
+In addition to the Jupyter notebooks and PDF notes, running this script also generates some example figures. We encourage students to look in the `scrips/` folder for some examples illustrating the theory. All PDF notes will also be made available on Ufora.
```
Comment (kirstvh) [inline README.md:47]: ✏️ scripts/ folder

**Pull request merged**